eth/tracers/native/opcode.go

```diff
@@ -0,0 +1,227 @@
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package native
+
+import (
+	"encoding/json"
+	"math/big"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/common/hexutil"
+	"github.com/ethereum/go-ethereum/common/math"
+	"github.com/ethereum/go-ethereum/core/vm"
+	"github.com/ethereum/go-ethereum/eth/tracers"
+	"github.com/ethereum/go-ethereum/eth/tracers/logger"
+)
+
+func init() {
+	tracers.DefaultDirectory.Register("opcodeTracer", newOpcodeTracer, false)
+}
+
+//go:generate go run github.com/fjl/gencodec -type CallFrameEnter -field-override callFrameEnterMarshaling -out gen_callframe_enter.go
+
+// CallFrameEnter is emitted every call frame entered.
+type CallFrameEnter struct {
+	Op    vm.OpCode      `json:"op"`
+	From  common.Address `json:"from"`
+	To    common.Address `json:"to"`
+	Input []byte         `json:"input"`
+	Gas   uint64         `json:"gas"`
+	Value *big.Int       `json:"value"`
+}
+
+// overrides for gencodec
+type callFrameEnterMarshaling struct {
+	Input  hexutil.Bytes
+	Gas    math.HexOrDecimal64
+	Value  *hexutil.Big
+	OpName string `json:"opName,omitempty"` // adds call to OpName() in MarshalJSON
+}
+
+// OpName formats the operand name in a human-readable format.
+func (c *CallFrameEnter) OpName() string {
+	return c.Op.String()
+}
+
+//go:generate go run github.com/fjl/gencodec -type CallFrameExit -field-override callFrameExitMarshaling -out gen_callframe_exit.go
+
+// CallFrameExit is emitted every call frame exits.
+type CallFrameExit struct {
+	From    common.Address `json:"from"`
+	To      common.Address `json:"to"`
+	Output  []byte         `json:"output"`
+	GasUsed uint64         `json:"gasUsed"`
+	Error   *string        `json:"error,omitempty"`
+}
+
+// overrides for gencodec
+type callFrameExitMarshaling struct {
+	Output  hexutil.Bytes
+	GasUsed math.HexOrDecimal64
+}
+
+type opcodeTracer struct {
+	noopTracer
+	traces    []json.RawMessage
+	callStack []common.Address
+	cfg       opcodeTracerConfig
+	env       *vm.EVM
+}
+
+type opcodeTracerConfig struct {
+	EnableMemory     bool // enable memory capture
+	DisableStack     bool // disable stack capture
+	EnableReturnData bool // enable return data capture
+	EnableCallFrames bool // enable call frame enter capture
+}
+
+// newCallTracer returns a native go tracer which tracks
+// call frames of a tx, and implements vm.EVMLogger.
+func newOpcodeTracer(ctx *tracers.Context, config json.RawMessage) (tracers.Tracer, error) {
+	var cfg opcodeTracerConfig
+	if config != nil {
+		if err := json.Unmarshal(config, &cfg); err != nil {
+			return nil, err
+		}
+	}
+	// First callframe contains tx context info
+	// and is populated on start and end.
+	return &opcodeTracer{traces: make([]json.RawMessage, 0), callStack: make([]common.Address, 0), cfg: cfg}, nil
+}
+
+func (t *opcodeTracer) appendTrace(item interface{}) {
```

Comment:
This can blow up the node quickly specially with memory enabled. We should rather stream out the result.